core/services/workflows/engine.go

```diff
@@ -815,26 +817,56 @@ func (e *Engine) workerForStepRequest(ctx context.Context, msg stepRequest) {
 	l := e.logger.With(platform.KeyStepRef, msg.stepRef, platform.KeyWorkflowExecutionID, msg.state.ExecutionID)
 	cma := e.cma.With(platform.KeyStepRef, msg.stepRef, platform.KeyWorkflowExecutionID, msg.state.ExecutionID)
 
+	curStepID := "UNSET"
+	curStep, verr := e.workflow.Vertex(msg.stepRef)
+	if verr == nil {
+		curStepID = curStep.ID
+	} else {
+		l.Errorf("failed to resolve step in workflow; error %v", verr)
+	}
+
+	// Fetch the workflow execution to get CreatedAt
+	ex, getErr := e.executionStates.Get(ctx, msg.state.ExecutionID)
+	if getErr != nil {
+		l.Errorf("failed to get execution state: %v", getErr)
+	}
+
 	l.Debug("executing on a step event")
 	stepState := &store.WorkflowExecutionStep{
 		Outputs:     store.StepOutput{},
 		ExecutionID: msg.state.ExecutionID,
 		Ref:         msg.stepRef,
 	}
 
+	// Check if the overall workflow execution has timed out
+	if ex.CreatedAt != nil && e.clock.Since(*ex.CreatedAt) > e.maxExecutionDuration {
```

Comment:
As discussed offline, I don't think this is necessary, as context cancellation should handle all timeouts.